src/components/migrated-bch.js

```diff
@@ -0,0 +1,324 @@
+        continue;
+      }
+      const chainPath = '/' + currentUnspent.chain + '/' + currentUnspent.index;
+      const subPath = signingKeychain.walletSubPath || '/0/0';
+      const path = signingKeychain.path + subPath + chainPath;
+      // derive the correct key
+      const privKey = hdPath.deriveKey(path);
+      const value = currentUnspent.value;
+
+      // do the signature flow
+      const subscript = new Buffer(currentUnspent.redeemScript, 'hex');
+      try {
+        txb.sign(inputIndex, privKey, subscript, sigHashType, value);
+      } catch (e) {
+        console.log(`got exception while signing unspent ${JSON.stringify(currentUnspent)}`);
+        console.trace(e);
+        throw e;
+      }
+
+      // now, let's verify the signature
+      transaction = txb.buildIncomplete();
+      const isSignatureVerified = bch.verifySignature(transaction, inputIndex, value);
+      if (!isSignatureVerified) {
+        throw new Error(`Could not verify signature on input #${inputIndex}`);
+      }
+    }
+
+    const tx = txb.buildIncomplete();
+    return {
+      hex: tx.toHex(),
+      id: tx.getId()
+    }
+  };
+
+  performRecovery = async () => {
+    const { bitgo } = this.props;
+    this.setState({ error: '', recovering: true });
+
+    const bch = bitgo.coin('bch');
+    const bchWallets = await bch.wallets().list();
+    const migratedWallet = _.find(bchWallets.wallets, w => w._wallet.migratedFrom === this.state.walletId);
+
+    if (!migratedWallet) {
+      throw new Error('could not find a bch wallet which was migrated from ' + this.state.walletId);
+    }
+
+    console.info('found bch wallet: ', migratedWallet.id());
+
+    let recoveryTx;
+    try {
+      recoveryTx = await this.createRecoveryTx(bch, migratedWallet);
+    } catch (e) {
+      if (e.message === 'insufficient balance') { // this is terribly unhelpful
+        e.message = 'Insufficient balance to recover';
+      }
+      this.collectLog(e.message);
+      this.setState({ error: e.message, recovering: false });
+    }
+
+    if (!recoveryTx || !recoveryTx.hex) {
+      console.error('Failed to create half-signed recovery transaction');
+      return;
+    }
+
+    let needsUnlock = false;
+    try {
+      await migratedWallet.submitTransaction({
+        txHex: recoveryTx.hex
+      });
+    } catch (e) {
+      if (e.message === 'needs unlock') {
+        // try again after unlocking
+        needsUnlock = true;
+      } else {
+        throw e;
+      }
+    }
+
+    if (needsUnlock) {
+      try {
+        await bitgo.unlock({ otp: this.state.twofa });
+        recoveryTx = await migratedWallet.submitTransaction({
+          txHex: recoveryTx.hex
+        });
+      } catch (e) {
+        // failed even after unlock - this is fatal
+        console.log('got error on submit after unlock');
+        console.error(e);
+        throw e;
+      }
+    }
+    this.setState({ recoveryTx, recovering: false });
+  };
+
+  saveTransaction = () => {
```

Comment:
We shouldn't need to save the transaction if we're submitting the transaction via the SDK. The other recoveries save transaction files so that they can be sent to support and cosigned manually.

Should we be submitting the transaction automatically? Or should we be saving a file for support?

Reply:
I'm open to suggestions here. I think auto-submit gives a better user experience and reduces support burden, but if you type in the wrong recovery address, you are in trouble.